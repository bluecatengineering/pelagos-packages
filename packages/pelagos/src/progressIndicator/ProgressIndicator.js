import {cloneElement} from 'react';
import PropTypes from 'prop-types';

import './ProgressIndicator.less';

/** Indicates progress on a multi-step process. */
const ProgressIndicator = ({className, current, children}) => (
	<ul className={`ProgressIndicator${className ? ` ${className}` : ''}`}>
		{children.map((child, index) =>
			cloneElement(child, {key: index, complete: index < current, current: current === index})
		)}
	</ul>
);

ProgressIndicator.propTypes = {
	/** The component class name(s). */
	className: PropTypes.string,
	/** The index of the current step. */
	current: PropTypes.number,
	/** The steps, must be instances of `ProgressStep`. */
	children: PropTypes.arrayOf(PropTypes.node),
};

export default ProgressIndicator;
