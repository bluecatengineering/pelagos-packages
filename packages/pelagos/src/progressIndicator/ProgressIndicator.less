@layer pelagos {
	.ProgressIndicator {
		display: flex;
		justify-content: stretch;
		list-style: none;
		padding: 0;
		margin: 0;
	}
}
