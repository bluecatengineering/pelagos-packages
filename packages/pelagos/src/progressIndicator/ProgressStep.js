import PropTypes from 'prop-types';
import {t} from '@bluecateng/l10n.macro';
import CheckmarkOutline from '@carbon/icons-react/es/CheckmarkOutline';
import CircleDash from '@carbon/icons-react/es/CircleDash';
import Incomplete from '@carbon/icons-react/es/Incomplete';
import Warning from '@carbon/icons-react/es/Warning';

import './ProgressStep.less';

const stepDescription = (complete, current, invalid) =>
	invalid ? t`Invalid` : current ? t`Current` : complete ? t`Complete` : t`Incomplete`;

/** Progress step. */
const ProgressStep = ({label, complete, current, invalid}) => (
	<li
		className={`ProgressStep ${
			complete ? 'ProgressStep--complete' : current ? 'ProgressStep--current' : 'ProgressStep--incomplete'
		}`}
		aria-current={current ? 'step' : undefined}>
		<span className="sr-only">, {stepDescription(complete, current, invalid)}</span>
		{invalid ? (
			<Warning className="ProgressStep__icon ProgressStep__icon--error" />
		) : current ? (
			<Incomplete className="ProgressStep__icon ProgressStep__icon--current" />
		) : complete ? (
			<CheckmarkOutline className="ProgressStep__icon ProgressStep__icon--complete" />
		) : (
			<CircleDash className="ProgressStep__icon ProgressStep__icon--incomplete" />
		)}
		<span className="ProgressStep__label">{label}</span>
	</li>
);

ProgressStep.propTypes = {
	/** The step label. */
	label: PropTypes.string,
	/** Whether the step is complete. */
	complete: PropTypes.bool,
	/** Whether the step the current one. */
	current: PropTypes.bool,
	/** Whether the step is not valid. */
	invalid: PropTypes.bool,
};

export default ProgressStep;
