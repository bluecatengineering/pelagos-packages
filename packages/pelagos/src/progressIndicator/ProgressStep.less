@import '../../less/spacing';

@layer pelagos {
	.ProgressStep {
		position: relative;
		display: flex;
		flex: 1;
		overflow: visible;
		border-top: 2px solid var(--border-subtle);
		align-items: center;
		min-width: 128px;
		cursor: default;

		&--complete,
		&--current {
			border-top-color: var(--border-interactive);
		}

		&__icon {
			min-width: 16px;
			height: 16px;
			margin: @sp-08 @sp-08 0 0;
		}

		&__icon--incomplete {
			color: var(--icon-primary);
		}

		&__icon--current,
		&__icon--complete {
			color: var(--interactive);
		}

		&__icon--error {
			color: var(--support-error);
		}

		&__label {
			margin-top: @sp-08;
		}
	}
}
