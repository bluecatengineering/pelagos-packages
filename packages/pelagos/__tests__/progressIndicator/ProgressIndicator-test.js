import {shallow} from 'enzyme';

import ProgressIndicator from '../../src/progressIndicator/ProgressIndicator';
import ProgressStep from '../../src/progressIndicator/ProgressStep';

jest.unmock('../../src/progressIndicator/ProgressIndicator');

describe('ProgressIndicator', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<ProgressIndicator current={1}>
					<ProgressStep />
					<ProgressStep />
				</ProgressIndicator>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when optional properties are set', () => {
			const wrapper = shallow(
				<ProgressIndicator className="TestClass" current={1}>
					<ProgressStep />
					<ProgressStep />
				</ProgressIndicator>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});
});
