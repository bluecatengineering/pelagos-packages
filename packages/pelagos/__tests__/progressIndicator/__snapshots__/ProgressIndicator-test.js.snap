// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressIndicator rendering renders expected elements 1`] = `
<ul
  className="ProgressIndicator"
>
  <ProgressStep
    complete={true}
    current={false}
    key={0}
  />
  <ProgressStep
    complete={false}
    current={true}
    key={1}
  />
</ul>
`;

exports[`ProgressIndicator rendering renders expected elements when optional properties are set 1`] = `
<ul
  className="ProgressIndicator TestClass"
>
  <ProgressStep
    complete={true}
    current={false}
    key={0}
  />
  <ProgressStep
    complete={false}
    current={true}
    key={1}
  />
</ul>
`;
